File: Lua/blumblumshub.lua
-- blumblumshub.lua
-- 
-- 

local rng = require("rng")

local defaultP, defaultQ = 3511, 17519

local blumblumshub = setmetatable({}, rng)
blumblumshub.__index = blumblumshub

local validatePrime, validateSeed
function blumblumshub.new(p, q, seed)
	if p and not q and not seed then
		p, q, seed = defaultP, defaultQ, p
	elseif not (p and q and seed) then
		error("blum-blum-shub requires either one or three parameters.")
	end
	
	if not validatePrime(p) or not validatePrime(q) then
		error("blum-blum-shub requires p and q to be primes congruent to 3 (mod 4).")
	end
	
	if not validateSeed(s) then
		error("blum-blum-shub requires the seed to be greater than 1.")
	end
	
	return { state = seed, pq = p * q }
end

function validatePrime(p)
	return p % 4 == 3
end

function validateSeed(seed)
	return seed ~= 0 and seed ~= 1
end

function blumblumshub:sample()
	local value = (self.state * self.state) % (self.pq)
	print("update, state = " .. self.state .. ", new state = " .. value)
	self.state = value
	return value / (self.pq - 1)
end

return blumblumshub

File: CSharp/RandomNumberGeneration/BlumBlumShub/BlumBlumShub.cs
using System;
using System.Diagnostics;
using RandomNumberGenerators.Utils.Primality;

namespace RandomNumberGenerators {
    public class BlumBlumShub : RandomNumberGenerator {
        private const long DefaultP = 3511;
        private const long DefaultQ = 17519;

        private readonly long pq;
        private long state;

        public BlumBlumShub(long seed)
            : this(DefaultP, DefaultQ, seed) { }

        public BlumBlumShub(long p, long q, long seed) {
            if(!ValidatePrime(p)) {
                throw new ArgumentOutOfRangeException(nameof(p), $"Blum-Blum-Shub requires {nameof(p)} to be a prime congruent to 3 (mod 4).");
            }

            if(!ValidatePrime(q)) {
                throw new ArgumentOutOfRangeException(nameof(q), $"Blum-Blum-Shub requires {nameof(q)} to be a prime congruent to 3 (mod 4).");
            }

            if(!ValidateSeed(seed)) {
                throw new ArgumentOutOfRangeException(nameof(seed), $"Blum-Blum-Shub requires {nameof(seed)} to be greater than 1.");
            }

            this.pq = p * q;
            this.state = seed;
        }

        private bool ValidatePrime(long candidate) {
            return new FermatPrimalityTest().Test(candidate) && candidate % 4 == 3;
        }

        private bool ValidateSeed(long seed) {
            return seed > 1;
        }

        public override double Sample() {
            var value = (this.state * this.state) % pq;
            Debug.WriteLine($"Update, State = {this.state}, New State = {value}");
            this.state = value;
            return (double)value / (pq - 1);
        }
    }
}
